fem/normal_deriv_restriction.hpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_NORMAL_DERIV_RESTRICTION
+#define MFEM_NORMAL_DERIV_RESTRICTION
+
+#include "../mesh/mesh.hpp"
+
+namespace mfem
+{
+
+class FiniteElementSpace;
+enum class ElementDofOrdering;
+
+/// @brief Class to compute face normal derivatives (in reference coordinate) of
+/// an L2 grid function (used internally by L2FaceRestriction).
+class L2NormalDerivativeFaceRestriction
+{
+protected:
+   const FiniteElementSpace &fes; ///< The L2 finite element space.
+   const FaceType face_type; ///< Face type: either boundary or interior.
+   const int dim; ///< Dimension of the mesh.
+   const int nf; ///< Number of faces of the given @a face_type.
+   const int ne; ///< Number of elements.
+   int ne_type; ///< Number of elements with faces of type face type
+
+   Array<int> face_to_elem; ///< Face-wise information array.
+   Array<int> elem_to_face; ///< Element-wise information array.
+   Array<int> face_to_vol; ///< maps face index to volume index
+
+public:
+   /// @brief Constructor.
+   /// @param[in] fes_ The associated FiniteElementSpace (should be L2/DG).
+   /// @param[in] ordering The ordering of L-DOFs.
+   /// @param[in] face_type_ Type of faces to compute restriction (interior or boundary).
+   L2NormalDerivativeFaceRestriction(const FiniteElementSpace &fes_,
+                                     const ElementDofOrdering ordering,
```

Comment:
The ordering argument comes from the `f_ordering` argument of the `L2FaceRestriction` constructor, so it has the same meaning:
```suggestion
   /// @param[in] f_ordering Request a specific face dof ordering
   /// @param[in] face_type_ Type of faces to compute restriction (interior or boundary).
   L2NormalDerivativeFaceRestriction(const FiniteElementSpace &fes_,
                                     const ElementDofOrdering f_ordering,
```